File: collect_polymarket_data.py
#!/usr/bin/env python3
"""
Polymarket Historical Data Collector (2025-2026)
Collects price history for all resolved markets from Jan 2025 to Feb 2026
"""

import requests
import sqlite3
import json
import time
from datetime import datetime, timezone
from typing import List, Dict, Optional
import sys

# API Configuration
BASE_URL = "https://gamma-api.polymarket.com"
RATE_LIMIT_DELAY = 0.5  # seconds between requests
MAX_RETRIES = 3

# Date range: Jan 1, 2025 to Feb 28, 2026
START_DATE = datetime(2025, 1, 1, tzinfo=timezone.utc)
END_DATE = datetime(2026, 2, 28, 23, 59, 59, tzinfo=timezone.utc)

class PolymarketCollector:
    def __init__(self, db_path: str):
        self.db_path = db_path
        self.session = requests.Session()
        self.stats = {
            'markets_found': 0,
            'markets_processed': 0,
            'price_points_collected': 0,
            'errors': [],
            'skipped_markets': []
        }
        self.init_database()
    
    def init_database(self):
        """Initialize SQLite database with schema"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Markets table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS markets (
                market_id TEXT PRIMARY KEY,
                question TEXT,
                end_date_iso TEXT,
                volume REAL,
                liquidity REAL,
                created_at TEXT,
                resolved_at TEXT,
                outcome TEXT
            )
        ''')
        
        # Price history table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS price_history (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                market_id TEXT NOT NULL,
                timestamp INTEGER NOT NULL,
                price REAL NOT NULL,
                volume REAL,
                FOREIGN KEY (market_id) REFERENCES markets(market_id)
            )
        ''')
        
        # Create indexes for efficient querying
        cursor.execute('''
            CREATE INDEX IF NOT EXISTS idx_price_market_time 
            ON price_history(market_id, timestamp)
        ''')
        
        conn.commit()
        conn.close()
        print(f"✓ Database initialized: {self.db_path}")
    
    def fetch_resolved_markets(self, limit: int = 100, offset: int = 0) -> List[Dict]:
        """Fetch resolved markets from Gamma API"""
        url = f"{BASE_URL}/markets"
        params = {
            'limit': limit,
            'offset': offset,
            'closed': 'true'  # Only resolved markets
        }
        
        for attempt in range(MAX_RETRIES):
            try:
                response = self.session.get(url, params=params, timeout=30)
                response.raise_for_status()
                time.sleep(RATE_LIMIT_DELAY)
                return response.json()
            except Exception as e:
                if attempt == MAX_RETRIES - 1:
                    self.stats['errors'].append(f"Failed to fetch markets (offset={offset}): {str(e)}")
                    return []
                time.sleep(2 ** attempt)  # Exponential backoff
        return []
    
    def is_in_date_range(self, market: Dict) -> bool:
        """Check if market resolved within our date range"""
        try:
            # Try different date fields
            end_date_str = market.get('endDate') or market.get('end_date_iso') or market.get('closed_time')
            if not end_date_str:
                return False
            
            # Parse ISO date
            if isinstance(end_date_str, str):
                end_date = datetime.fromisoformat(end_date_str.replace('Z', '+00:00'))
            else:
                # Unix timestamp
                end_date = datetime.fromtimestamp(int(end_date_str), tz=timezone.utc)
            
            return START_DATE <= end_date <= END_DATE
        except Exception as e:
            return False
    
    def fetch_price_history(self, market_id: str) -> Optional[List[Dict]]:
        """Fetch price history for a specific market"""
        # Try different possible endpoints
        endpoints = [
            f"{BASE_URL}/prices-history",
            f"{BASE_URL}/markets/{market_id}/prices-history",
            f"{BASE_URL}/prices",
        ]
        
        for endpoint in endpoints:
            for attempt in range(MAX_RETRIES):
                try:
                    params = {'market': market_id} if 'prices-history' in endpoint else {}
                    response = self.session.get(endpoint, params=params, timeout=30)
                    
                    if response.status_code == 200:
                        time.sleep(RATE_LIMIT_DELAY)
                        data = response.json()
                        # Handle different response formats
                        if isinstance(data, list):
                            return data
                        elif isinstance(data, dict) and 'history' in data:
                            return data['history']
                        elif isinstance(data, dict) and 'data' in data:
                            return data['data']
                        return []
                    elif response.status_code == 404:
                        break  # Try next endpoint
                    
                    time.sleep(2 ** attempt)
                except Exception as e:
                    if attempt == MAX_RETRIES - 1:
                        continue  # Try next endpoint
                    time.sleep(2 ** attempt)
        
        return None
    
    def store_market(self, market: Dict):
        """Store market metadata in database"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        try:
            cursor.execute('''
                INSERT OR REPLACE INTO markets 
                (market_id, question, end_date_iso, volume, liquidity, created_at, resolved_at, outcome)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            ''', (
                market.get('id') or market.get('condition_id'),
                market.get('question') or market.get('description'),
                market.get('endDate') or market.get('end_date_iso'),
                float(market.get('volume', 0) or 0),
                float(market.get('liquidity', 0) or 0),
                market.get('createdAt') or market.get('created_at'),
                market.get('resolvedAt') or market.get('resolved_at'),
                market.get('outcome')
            ))
            conn.commit()
        finally:
            conn.close()
    
    def store_price_history(self, market_id: str, history: List[Dict]):
        """Store price history in database"""
        if not history:
            return
        
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        try:
            for point in history:
                # Extract timestamp and price from different possible formats
                timestamp = point.get('t') or point.get('timestamp') or point.get('time')
                price = point.get('p') or point.get('price')
                volume = point.get('v') or point.get('volume')
                
                if timestamp and price is not None:
                    cursor.execute('''
                        INSERT INTO price_history (market_id, timestamp, price, volume)
                        VALUES (?, ?, ?, ?)
                    ''', (market_id, int(timestamp), float(price), float(volume) if volume else None))
            
            conn.commit()
            self.stats['price_points_collected'] += len(history)
        finally:
            conn.close()
    
    def collect_all_data(self):
        """Main collection loop"""
        print("🚀 Starting Polymarket data collection (2025-2026)")
        print(f"📅 Date range: {START_DATE.date()} to {END_DATE.date()}\n")
        
        offset = 0
        limit = 100
        markets_in_range = []
        
        # Step 1: Collect all markets in date range
        print("📊 Phase 1: Discovering resolved markets...")
        while True:
            markets = self.fetch_resolved_markets(limit=limit, offset=offset)
            if not markets:
                break
            
            for market in markets:
                if self.is_in_date_range(market):
                    markets_in_range.append(market)
                    self.stats['markets_found'] += 1
            
            print(f"  Scanned {offset + len(markets)} markets, found {len(markets_in_range)} in range...", end='\r')
            
            if len(markets) < limit:
                break
            
            offset += limit
            
            # Safety limit to avoid infinite loops
            if offset > 10000:
                print("\n⚠️  Reached safety limit of 10,000 markets")
                break
        
        print(f"\n✓ Found {len(markets_in_range)} markets in date range")
        
        # Step 2: Sort by volume (high-volume first)
        markets_in_range.sort(key=lambda m: float(m.get('volume', 0) or 0), reverse=True)
        
        # Step 3: Collect price history
        print(f"\n📈 Phase 2: Collecting price history...")
        for i, market in enumerate(markets_in_range, 1):
            market_id = market.get('id') or market.get('condition_id')
            question = market.get('question', 'Unknown')[:60]
            volume = market.get('volume', 0)
            
            print(f"  [{i}/{len(markets_in_range)}] {question}... (vol: ${volume:,.0f})")
            
            # Store market metadata
            self.store_market(market)
            
            # Fetch and store price history
            history = self.fetch_price_history(market_id)
            if history:
                self.store_price_history(market_id, history)
                print(f"    ✓ Collected {len(history)} price points")
            else:
                self.stats['skipped_markets'].append({
                    'id': market_id,
                    'question': question,
                    'reason': 'No price history available'
                })
                print(f"    ⚠️  No price history found")
            
            self.stats['markets_processed'] += 1
        
        print("\n✅ Collection complete!")
        self.print_stats()
    
    def print_stats(self):
        """Print collection statistics"""
        print("\n" + "="*60)
        print("📊 COLLECTION STATISTICS")
        print("="*60)
        print(f"Markets found in range:     {self.stats['markets_found']}")
        print(f"Markets processed:          {self.stats['markets_processed']}")
        print(f"Price points collected:     {self.stats['price_points_collected']:,}")
        print(f"Markets skipped:            {len(self.stats['skipped_markets'])}")
        print(f"Errors encountered:         {len(self.stats['errors'])}")
        print("="*60)
    
    def generate_quality_report(self, report_path: str):
        """Generate data quality report"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Gather statistics
        cursor.execute("SELECT COUNT(*) FROM markets")
        total_markets = cursor.fetchone()[0]
        
        cursor.execute("SELECT COUNT(*) FROM price_history")
        total_prices = cursor.fetchone()[0]
        
        cursor.execute("""
            SELECT m.market_id, m.question, COUNT(ph.id) as price_count
            FROM markets m
            LEFT JOIN price_history ph ON m.market_id = ph.market_id
            GROUP BY m.market_id
            HAVING price_count = 0
        """)
        markets_without_prices = cursor.fetchall()
        
        cursor.execute("""
            SELECT market_id, COUNT(*) as count
            FROM price_history
            GROUP BY market_id
            ORDER BY count ASC
            LIMIT 10
        """)
        sparse_markets = cursor.fetchall()
        
        conn.close()
        
        # Write report
        with open(report_path, 'w') as f:
            f.write("# Data Quality Report: Polymarket Historical Prices (2025-2026)\n\n")
            f.write(f"**Generated:** {datetime.now().isoformat()}\n\n")
            
            f.write("## Summary\n\n")
            f.write(f"- **Total Markets:** {total_markets}\n")
            f.write(f"- **Total Price Points:** {total_prices:,}\n")
            f.write(f"- **Average Price Points per Market:** {total_prices/total_markets if total_markets > 0 else 0:.1f}\n")
            f.write(f"- **Markets Without Price Data:** {len(markets_without_prices)}\n\n")
            
            f.write("## Data Gaps\n\n")
            if markets_without_prices:
                f.write("### Markets Missing Price History\n\n")
                for market_id, question, _ in markets_without_prices[:20]:
                    f.write(f"- `{market_id}`: {question}\n")
                if len(markets_without_prices) > 20:
                    f.write(f"\n*... and {len(markets_without_prices) - 20} more*\n")
            else:
                f.write("✅ All markets have price history data.\n")
            
            f.write("\n### Markets with Sparse Data (<50 points)\n\n")
            for market_id, count in sparse_markets:
                f.write(f"- `{market_id}`: {count} price points\n")
            
            if self.stats['errors']:
                f.write("\n## Errors Encountered\n\n")
                for error in self.stats['errors']:
                    f.write(f"- {error}\n")
            
            if self.stats['skipped_markets']:
                f.write("\n## Skipped Markets\n\n")
                for skip in self.stats['skipped_markets'][:50]:
                    f.write(f"- `{skip['id']}`: {skip['question']} - {skip['reason']}\n")
                if len(self.stats['skipped_markets']) > 50:
                    f.write(f"\n*... and {len(self.stats['skipped_markets']) - 50} more*\n")
            
            f.write("\n## Notes\n\n")
            f.write("- Data source: Polymarket Gamma API\n")
            f.write("- Only resolved markets included\n")
            f.write("- Markets sorted by volume (high to low) during collection\n")
            f.write("- Rate limiting: 0.5s delay between requests\n")
        
        print(f"✓ Quality report generated: {report_path}")


if __name__ == "__main__":
    db_path = "historical_2025_2026.db"
    report_path = "DATA_QUALITY_2025_2026.md"
    
    collector = PolymarketCollector(db_path)
    
    try:
        collector.collect_all_data()
        collector.generate_quality_report(report_path)
        
        print(f"\n✅ SUCCESS!")
        print(f"   Database: {db_path}")
        print(f"   Report:   {report_path}")
        
    except KeyboardInterrupt:
        print("\n\n⚠️  Collection interrupted by user")
        collector.generate_quality_report(report_path)
        sys.exit(1)
    except Exception as e:
        print(f"\n\n❌ Error: {str(e)}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

File: collect_polymarket_data.js
#!/usr/bin/env node
/**
 * Polymarket Historical Data Collector (2025-2026)
 * Collects price history for all resolved markets from Jan 2025 to Feb 2026
 */

const https = require('https');
const fs = require('fs');
const sqlite3 = require('better-sqlite3');

// API Configuration
const BASE_URL = 'gamma-api.polymarket.com';
const RATE_LIMIT_DELAY = 500; // milliseconds
const MAX_RETRIES = 3;

// Date range: Jan 1, 2025 to Feb 28, 2026
const START_DATE = new Date('2025-01-01T00:00:00Z');
const END_DATE = new Date('2026-02-28T23:59:59Z');

class PolymarketCollector {
  constructor(dbPath) {
    this.dbPath = dbPath;
    this.stats = {
      markets_found: 0,
      markets_processed: 0,
      price_points_collected: 0,
      errors: [],
      skipped_markets: []
    };
    this.initDatabase();
  }

  initDatabase() {
    this.db = new sqlite3(this.dbPath);
    
    // Markets table
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS markets (
        market_id TEXT PRIMARY KEY,
        question TEXT,
        end_date_iso TEXT,
        volume REAL,
        liquidity REAL,
        created_at TEXT,
        resolved_at TEXT,
        outcome TEXT
      )
    `);
    
    // Price history table
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS price_history (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        market_id TEXT NOT NULL,
        timestamp INTEGER NOT NULL,
        price REAL NOT NULL,
        volume REAL,
        FOREIGN KEY (market_id) REFERENCES markets(market_id)
      )
    `);
    
    // Create indexes
    this.db.exec(`
      CREATE INDEX IF NOT EXISTS idx_price_market_time 
      ON price_history(market_id, timestamp)
    `);
    
    console.log(`✓ Database initialized: ${this.dbPath}`);
  }

  async httpGet(path, params = {}) {
    return new Promise((resolve, reject) => {
      const queryString = Object.keys(params)
        .map(key => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`)
        .join('&');
      
      const fullPath = queryString ? `${path}?${queryString}` : path;
      
      const options = {
        hostname: BASE_URL,
        path: fullPath,
        method: 'GET',
        headers: {
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }
      };

      const req = https.request(options, (res) => {
        let data = '';
        
        res.on('data', (chunk) => {
          data += chunk;
        });
        
        res.on('end', () => {
          if (res.statusCode === 200) {
            try {
              resolve(JSON.parse(data));
            } catch (e) {
              reject(new Error(`JSON parse error: ${e.message}`));
            }
          } else {
            reject(new Error(`HTTP ${res.statusCode}: ${data}`));
          }
        });
      });

      req.on('error', reject);
      req.setTimeout(30000, () => {
        req.destroy();
        reject(new Error('Request timeout'));
      });
      
      req.end();
    });
  }

  async sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  async fetchResolvedMarkets(limit = 100, offset = 0) {
    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
      try {
        const data = await this.httpGet('/markets', {
          limit,
          offset,
          closed: 'true'
        });
        await this.sleep(RATE_LIMIT_DELAY);
        return Array.isArray(data) ? data : [];
      } catch (e) {
        if (attempt === MAX_RETRIES - 1) {
          this.stats.errors.push(`Failed to fetch markets (offset=${offset}): ${e.message}`);
          return [];
        }
        await this.sleep(Math.pow(2, attempt) * 1000);
      }
    }
    return [];
  }

  isInDateRange(market) {
    try {
      const endDateStr = market.endDate || market.end_date_iso || market.closed_time;
      if (!endDateStr) return false;
      
      const endDate = new Date(endDateStr);
      return endDate >= START_DATE && endDate <= END_DATE;
    } catch (e) {
      return false;
    }
  }

  async fetchPriceHistory(marketId) {
    const endpoints = [
      { path: '/prices-history', params: { market: marketId } },
      { path: `/markets/${marketId}/prices-history`, params: {} },
      { path: '/prices', params: { market: marketId } }
    ];
    
    for (const endpoint of endpoints) {
      for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
        try {
          const data = await this.httpGet(endpoint.path, endpoint.params);
          await this.sleep(RATE_LIMIT_DELAY);
          
          if (Array.isArray(data)) return data;
          if (data && data.history) return data.history;
          if (data && data.data) return data.data;
          return [];
        } catch (e) {
          if (e.message.includes('404')) break; // Try next endpoint
          if (attempt === MAX_RETRIES - 1) continue;
          await this.sleep(Math.pow(2, attempt) * 1000);
        }
      }
    }
    return null;
  }

  storeMarket(market) {
    const stmt = this.db.prepare(`
      INSERT OR REPLACE INTO markets 
      (market_id, question, end_date_iso, volume, liquidity, created_at, resolved_at, outcome)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?)
    `);
    
    stmt.run(
      market.id || market.condition_id,
      market.question || market.description,
      market.endDate || market.end_date_iso,
      parseFloat(market.volume || 0),
      parseFloat(market.liquidity || 0),
      market.createdAt || market.created_at,
      market.resolvedAt || market.resolved_at,
      market.outcome
    );
  }

  storePriceHistory(marketId, history) {
    if (!history || history.length === 0) return;
    
    const stmt = this.db.prepare(`
      INSERT INTO price_history (market_id, timestamp, price, volume)
      VALUES (?, ?, ?, ?)
    `);
    
    const insert = this.db.transaction((points) => {
      for (const point of points) {
        const timestamp = point.t || point.timestamp || point.time;
        const price = point.p || point.price;
        const volume = point.v || point.volume;
        
        if (timestamp && price !== undefined) {
          stmt.run(marketId, parseInt(timestamp), parseFloat(price), volume ? parseFloat(volume) : null);
        }
      }
    });
    
    insert(history);
    this.stats.price_points_collected += history.length;
  }

  async collectAllData() {
    console.log('🚀 Starting Polymarket data collection (2025-2026)');
    console.log(`📅 Date range: ${START_DATE.toISOString().split('T')[0]} to ${END_DATE.toISOString().split('T')[0]}\n`);
    
    let offset = 0;
    const limit = 100;
    const marketsInRange = [];
    
    // Phase 1: Discover markets
    console.log('📊 Phase 1: Discovering resolved markets...');
    while (true) {
      const markets = await this.fetchResolvedMarkets(limit, offset);
      if (!markets || markets.length === 0) break;
      
      for (const market of markets) {
        if (this.isInDateRange(market)) {
          marketsInRange.push(market);
          this.stats.markets_found++;
        }
      }
      
      process.stdout.write(`  Scanned ${offset + markets.length} markets, found ${marketsInRange.length} in range...\r`);
      
      if (markets.length < limit) break;
      offset += limit;
      
      if (offset > 10000) {
        console.log('\n⚠️  Reached safety limit of 10,000 markets');
        break;
      }
    }
    
    console.log(`\n✓ Found ${marketsInRange.length} markets in date range`);
    
    // Sort by volume (high to low)
    marketsInRange.sort((a, b) => parseFloat(b.volume || 0) - parseFloat(a.volume || 0));
    
    // Phase 2: Collect price history
    console.log(`\n📈 Phase 2: Collecting price history...`);
    for (let i = 0; i < marketsInRange.length; i++) {
      const market = marketsInRange[i];
      const marketId = market.id || market.condition_id;
      const question = (market.question || 'Unknown').substring(0, 60);
      const volume = market.volume || 0;
      
      console.log(`  [${i + 1}/${marketsInRange.length}] ${question}... (vol: $${parseFloat(volume).toLocaleString()})`);
      
      // Store market metadata
      this.storeMarket(market);
      
      // Fetch and store price history
      const history = await this.fetchPriceHistory(marketId);
      if (history && history.length > 0) {
        this.storePriceHistory(marketId, history);
        console.log(`    ✓ Collected ${history.length} price points`);
      } else {
        this.stats.skipped_markets.push({
          id: marketId,
          question,
          reason: 'No price history available'
        });
        console.log(`    ⚠️  No price history found`);
      }
      
      this.stats.markets_processed++;
    }
    
    console.log('\n✅ Collection complete!');
    this.printStats();
  }

  printStats() {
    console.log('\n' + '='.repeat(60));
    console.log('📊 COLLECTION STATISTICS');
    console.log('='.repeat(60));
    console.log(`Markets found in range:     ${this.stats.markets_found}`);
    console.log(`Markets processed:          ${this.stats.markets_processed}`);
    console.log(`Price points collected:     ${this.stats.price_points_collected.toLocaleString()}`);
    console.log(`Markets skipped:            ${this.stats.skipped_markets.length}`);
    console.log(`Errors encountered:         ${this.stats.errors.length}`);
    console.log('='.repeat(60));
  }

  generateQualityReport(reportPath) {
    // Gather statistics
    const totalMarkets = this.db.prepare('SELECT COUNT(*) as count FROM markets').get().count;
    const totalPrices = this.db.prepare('SELECT COUNT(*) as count FROM price_history').get().count;
    
    const marketsWithoutPrices = this.db.prepare(`
      SELECT m.market_id, m.question, COUNT(ph.id) as price_count
      FROM markets m
      LEFT JOIN price_history ph ON m.market_id = ph.market_id
      GROUP BY m.market_id
      HAVING price_count = 0
    `).all();
    
    const sparseMarkets = this.db.prepare(`
      SELECT market_id, COUNT(*) as count
      FROM price_history
      GROUP BY market_id
      ORDER BY count ASC
      LIMIT 10
    `).all();
    
    // Write report
    let report = '# Data Quality Report: Polymarket Historical Prices (2025-2026)\n\n';
    report += `**Generated:** ${new Date().toISOString()}\n\n`;
    
    report += '## Summary\n\n';
    report += `- **Total Markets:** ${totalMarkets}\n`;
    report += `- **Total Price Points:** ${totalPrices.toLocaleString()}\n`;
    report += `- **Average Price Points per Market:** ${totalMarkets > 0 ? (totalPrices / totalMarkets).toFixed(1) : 0}\n`;
    report += `- **Markets Without Price Data:** ${marketsWithoutPrices.length}\n\n`;
    
    report += '## Data Gaps\n\n';
    if (marketsWithoutPrices.length > 0) {
      report += '### Markets Missing Price History\n\n';
      for (let i = 0; i < Math.min(20, marketsWithoutPrices.length); i++) {
        const m = marketsWithoutPrices[i];
        report += `- \`${m.market_id}\`: ${m.question}\n`;
      }
      if (marketsWithoutPrices.length > 20) {
        report += `\n*... and ${marketsWithoutPrices.length - 20} more*\n`;
      }
    } else {
      report += '✅ All markets have price history data.\n';
    }
    
    report += '\n### Markets with Sparse Data (<50 points)\n\n';
    for (const m of sparseMarkets) {
      report += `- \`${m.market_id}\`: ${m.count} price points\n`;
    }
    
    if (this.stats.errors.length > 0) {
      report += '\n## Errors Encountered\n\n';
      for (const error of this.stats.errors) {
        report += `- ${error}\n`;
      }
    }
    
    if (this.stats.skipped_markets.length > 0) {
      report += '\n## Skipped Markets\n\n';
      for (let i = 0; i < Math.min(50, this.stats.skipped_markets.length); i++) {
        const skip = this.stats.skipped_markets[i];
        report += `- \`${skip.id}\`: ${skip.question} - ${skip.reason}\n`;
      }
      if (this.stats.skipped_markets.length > 50) {
        report += `\n*... and ${this.stats.skipped_markets.length - 50} more*\n`;
      }
    }
    
    report += '\n## Notes\n\n';
    report += '- Data source: Polymarket Gamma API\n';
    report += '- Only resolved markets included\n';
    report += '- Markets sorted by volume (high to low) during collection\n';
    report += '- Rate limiting: 500ms delay between requests\n';
    
    fs.writeFileSync(reportPath, report);
    console.log(`✓ Quality report generated: ${reportPath}`);
  }

  close() {
    this.db.close();
  }
}

// Main execution
(async () => {
  const dbPath = 'historical_2025_2026.db';
  const reportPath = 'DATA_QUALITY_2025_2026.md';
  
  const collector = new PolymarketCollector(dbPath);
  
  try {
    await collector.collectAllData();
    collector.generateQualityReport(reportPath);
    collector.close();
    
    console.log(`\n✅ SUCCESS!`);
    console.log(`   Database: ${dbPath}`);
    console.log(`   Report:   ${reportPath}`);
  } catch (e) {
    console.error(`\n❌ Error: ${e.message}`);
    console.error(e.stack);
    collector.close();
    process.exit(1);
  }
})();
